### PR TITLE
Rewrite scalar dumper for correctness

### LINKED ISSUE #238: Dumper must not add trailing line break at the end
Dumper always adds one trailing line break at the end. It was a feature before we added block scalars support, but now it break dumps in cases, when last object in the stream is `|+` or `>+` scalar.

### PR DESCRIPTION
**Pull request by aepsilon**
Rewrote the scalar dumper with an emphasis on rigor, based on a
careful reading of the YAML 1.2 spec and testing against LibYAML.
### Corrections

Block style chomping accounts for the extra newline
added by subsequent elements or at the end.
Fixes #238.

Top-level block scalars are indented.
Fixes dumper side of #221.

Folded style should be reliable now. The old dumper was not aware of
the subtle rules for more-indented lines.
More-indented lines no longer cause extra newlines to be added.
(The old dumper also had an off-by-one error that dropped
the last character from a long first word on a line.)
Fixes #215, #222. Removes #217 workaround.

Literal and folded styles can represent arbitrary "printable" strings,
including leading/trailing whitespace.
(Hopefully this improves readability for many use cases.)

NB. For the 'construct-string-types' dumper test, the strings are
dumped correctly. The test fails because a loader bug
loses a newline when using the keep "+" chomping indicator.
### Additional improvements

Added test suite covering several former bugs, new edge cases,
and new expected behaviors. All previous applicable tests are included.

Styles are chosen in a well-defined order.
- Plain and single are preferred for single lines under the width limit.
- Folded is preferred when a line is longer than the width limit
  (if enabled) and does not begin with a space.
- Literal is used for all other multi-line "printable" strings.
- Double-quoted is only used when a string is unrepresentable otherwise.

Plain style is relaxed to include more characters and strings.
Progress on #183.

As indent increases, line width decreases down to min(lineWidth, 40).
Setting lineWidth to -1 disables wrapping.
Previously there was unexpected behavior past 40 indent:
narrow lineWidth suddenly jumped to 40,
and -1 (no limit) suddenly enforced a limit of 40.

### CODE REVIEW

File: test/issues/0217.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ var assert = require('assert');
 var yaml   = require('../../');
 
 
-test('Javascript functions are not folded', function () {
+test.skip('Javascript functions are not folded', function () {
```
Comment (puzrin) [inline test/issues/0217.js]: Why are those tests skipped?

Comment (aepsilon) [inline test/issues/0217.js]: [As @dervus expected](https://github.com/nodeca/js-yaml/pull/215#issuecomment-150790606), the actual problem was the folded style dumper was buggy. It tried to break more-indented lines, which resulted in newlines being added. Folding more-indented lines is not allowed in YAML.

Now that the folder is fixed it should work for all string data regardless of content.
This line tests for a workaround that doesn't fix the root problem (see #222).

Although I think it's better that the root problem is fixed, I can add the workaround back in if desired.

Comment (puzrin) [inline test/issues/0217.js]: I mean, tests were added to check that fns & regexes are not broken. I don't mean tests should stay intact, but those should continue to work. There are no need to disable those cimpletely, for historic reasons.

Comment (aepsilon) [inline test/issues/0217.js]: Naturally, we want to make sure they continue to work.

The dumper doesn't differentiate between tags anymore. The current test demands no folded style, even though the data is dumped correctly. Keeping the test as-is means putting the workaround back in the dumper.

I could change the test to dump then load several examples of functions and regexes.

Comment (puzrin) [inline test/issues/0217.js]: Yes, update tests to reflect your changes. No workaround in dumper needed.

File: test/units/dump-scalar-styles.js [hunk 2]
```diff
@@ -3,79 +3,234 @@
...
+  suite('Folded style', function () {
+    (function () {
+      var content = (function () {
+        var result = '';
+        var i = 1000;
+        for (var para = 1; para <= 7; para++) {
+          result += '\n';
+          // indent paragraphs 3 and 4
+          if (para === 3 || para === 4) {
+            result += ' '.repeat(para);
```
Comment (puzrin) [inline test/units/dump-scalar-styles.js]: `.repeat()` is poorly supported. Needs helper or `Array(para).join(' ')` to fix tests.

Run tests under node 0.10 for sure.

**Comment by puzrin** [comment]
Very impressing. This week surprised me so much - 2 PRs (here & in `pako`), impossible without deep understanding of subject by authors.

@dervus could you take a look?

**Comment by dervus** [comment]
I can take a look on the weekend. Very busy right now, sorry.

**Comment by aepsilon** [comment]
Thanks for the compliment and the quick responses. I've updated the tests.

Regarding the loader bug, it seems to occur whenever the scalar is anything but trailing empty lines:

    > yaml.safeLoad('a: |+\n\n\n')
    { a: '\n\n' }
    > yaml.safeLoad('a: |+\n  \n\n')
    { a: '\n\n' }
    > yaml.safeLoad('a: |+\n b\n\n')
    { a: 'b\n' }

    > yaml.safeLoad('a: >+\n\n\n')
    { a: '\n\n' }
    > yaml.safeLoad('a: >+\n  \n\n')
    { a: '\n\n' }
    > yaml.safeLoad('a: >+\n b\n\n')
    { a: 'b\n' }

File: lib/js-yaml/dumper.js [hunk 3]
```diff
@@ -12,7 +12,7 @@ var _hasOwnProperty = Object.prototype.hasOwnProperty;
 
 var CHAR_TAB                  = 0x09; /* Tab */
 var CHAR_LINE_FEED            = 0x0A; /* LF */
-var CHAR_CARRIAGE_RETURN      = 0x0D; /* CR */
+// var CHAR_CARRIAGE_RETURN      = 0x0D; /* CR */ // unused
```
Comment (dervus) [inline lib/js-yaml/dumper.js]: No reason to keep it in any form here if it's not in use. It can be just re-added in future if needed.

File: test/issues/0217.js [hunk 4]
```diff
@@ -5,19 +5,107 @@ var assert = require('assert');
...
+    'function go(n) { result.push(n); return (n === 1) ? result : go(n % 2 === 0  ?  n / 2  :  3 * n + 1); }\n' +
+    'return go(start >= 1 ? Math.floor(start) : 1);\n' +
+    '}');
+
+  var fnRot13 = escapeFnString(
+    // single-line function.
+    // note the "{return" is so the line doesn't start with a space.
+    'function rot13(s) {return String.fromCharCode.apply(null, s.split("")' +
+    '.map(function (c) { return ((c.toLowerCase().charCodeAt(0) - 97) + 13) % 26 + 97; })); }'
+  );
```
Comment (dervus) [inline test/issues/0217.js]: I think these functions and constants should be defined at the top-level as two other helper functions.

Comment (dervus) [inline test/issues/0217.js]: I understand the logic though, so it doesn't matter.

**Comment by dervus** [comment]
Also `assert.equal` in tests could be replaced with `assert.strictEqual` for consistency. But I don't really care about it.

Overall that's very good piece of work. Thank you, @aepsilon!

@puzrin, it's good for merge. However, we need to fix that loader bug before releasing this. I'll try to look into it during next week.

**Comment by puzrin** [comment]
https://github.com/nodeca/js-yaml/commit/e0c4e4e3dea07b68bd62022a38944ce3a5b1afb9 - updated asserts use everywhere.
